### ISSUE TITLE
Autocomplete form fields

### ISSUE THREAD

**Issue by neon-ninja** [swcarpentry/amy#233]
Would Amy benefit from installing ajax autocomplete for specific form fields - for example, when selecting a user for a new task.

https://pypi.python.org/pypi/django-ajax-selects
https://github.com/crucialfelix/django-ajax-selects/

**Comment by wking** [comment]
On Thu, Mar 19, 2015 at 08:39:33PM -0700, Nick Young wrote:
> Would Amy benefit from installing ajax autocomplete for specific
> form fields - for example, when selecting a user for a new task.

That sounds reasonable.  How does task-selection work now?  There are
few enough roles that selecting the role with Django's stock select
should work fine, but picking a from an event page or an event from a
person page would probably benefit from this approach.  I'd wait for
more folks to chime in before diving into the implementation though.

**Comment by gvwilson** [comment]
+1 to doing this now for fields with a small number of fixed values 
(tasks, badges) - we can extend to larger things (users, events) later.

**Comment by shapiromatron** [comment]
I've worked a good-bit with [django-selectable](https://github.com/mlavin/django-selectable), and find it works really well and is simple to integrate. Used it extensively with `django-crispy-forms` as well so I don't think there are any conflicts. Very few changes to the html templates; otherwise it's as simple as customizing the widgets used in the forms, and creating "lookups" which define an AJAX request and a queryset. 

Docs: http://django-selectable.readthedocs.org/en/v0.9.X/quick-start.html

I would generally implement in any cases where there are FK or M2M fields with greater than 30 or so options; these are the fields I identified from just scanning the source

- InstructorsForm.airport 
- Person.airport 
- Event.tags, Event.site in admin
- Task in admin (event, person, likely not role as there aren't too many options)

Want me to give it a go?

**Comment by pbanaszkiewicz** [comment]
Hey @shapiromatron!

django-selectable looks really nice and I'd love to see it in Amy. However, I don't want you to start working on it yet, because we have #234 and #242 blocking (and I'd rather you avoided unnecessary rebasing).

By the way, do you perhaps have any links on how to use django-selectable and django-crispy-forms and class based views altogether? I hope we won't have to write a ModelForm for every Create and UpdateView we have.

Cheers,
Piotr

**Comment by shapiromatron** [comment]
Great, I'll wait for you to merge before starting. 

Here's an example from my own project of how to integrate, using the `EffectTag` as an example. My project is overly complicated as I was learning as I went, but you can see the general steps with this example: 

1. Add lookup(s) to register an AJAX search for a model or model-field:
https://github.com/shapiromatron/hawc/blob/master/project/assessment/lookups.py#L12

2. Override original field on forms.py to make a selectable field:
https://github.com/shapiromatron/hawc/blob/master/project/assessment/forms.py#L117

3. In the template, add a few html-tags and requirements:
https://github.com/shapiromatron/hawc/blob/master/project/templates/assessment/effecttag_form.html#L4

The AJAX URLs are automatically-created when adding the package to settings, so no mods to views.py or urls.py are required.

**Comment by shapiromatron** [comment]
But to answer your question, yes, in order to use the autocomplete/selectable magic, we'll likely need to add some additional ModelForms.

**Comment by pbanaszkiewicz** [comment]
Hey @shapiromatron,

#234 and #242 are now merged, so you can start working on a better autocomplete features. Thanks!
